### PR TITLE
[RateLimiter] Adding SlidingWindow algorithm

### PR DESCRIPTION
**Pull request by Nyholm**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | 
| Tickets       | 
| License       | MIT
| Doc PR        | Should be added

This is a draft PR to make sure we like the idea. I'll keep working on adding tests.

### CODE REVIEW

**Comment by wouterj** [comment]
:+1: I like this. Thanks!

File: src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        $limiter->consume(8);
+        sleep(75);
+
+        $limit = $limiter->consume();
+        $this->assertTrue($limit->isAccepted());
+
+        // We are 25% into the new window
+        $limit = $limiter->consume(5);
+        $this->assertFalse($limit->isAccepted());
```
Comment (wouterj) [inline src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php:45]: Maybe also add `$this->assertTrue($limiter->consume(4)->isAccepted());`? (if I understand this test correctly, that should be the case here)

File: src/Symfony/Component/RateLimiter/SlidingWindow.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Store for the rest of this time frame and next.
+     */
+    public function getExpirationTime(): ?int
+    {
+        if ($this->cached) {
+            return null;
+        }
```
Comment (wouterj) [inline src/Symfony/Component/RateLimiter/SlidingWindow.php:95]: Isn't `$cached` always false? Given that we serialize the whole object and set it to `false` in the constructor?

Looking at the code, I would also expect to just always return `2 * $this->intervalInSeconds` here.

File: src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php (hunk 1 above)
Comment (wouterj) [reply src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php:45]: Or `$this->assertEquals(4, $limit->getRemainingTokens());`

**Review by wouterj — APPROVED** [review]
Apart from my comments, this looks good to me.

@fabpot would be cool to also have this in 5.2, though it's not 100% needed. Sliding windows generally is better than fixed windows, as they flatten bursts much better.

File: src/Symfony/Component/RateLimiter/SlidingWindow.php (hunk 2 above)
Comment (Nyholm) [reply src/Symfony/Component/RateLimiter/SlidingWindow.php:95]: You are correct. 

I'll add back the sleep method

File: src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php (hunk 1 above)
Comment (Nyholm) [reply src/Symfony/Component/RateLimiter/Tests/SlidingWindowLimiterTest.php:45]: Thank you. I've added that. 

I also replaced "created at" to "window end at". It makes things simpler.

File: src/Symfony/Component/RateLimiter/SlidingWindowLimiter.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            $window->add($tokens);
+            $this->storage->save($window);
+
+            return new Limit($this->getAvailableTokens($window->getHitCount()), $window->getRetryAfter(), true);
+        } finally {
+            $this->lock->release();
+        }
+    }
+
+    private function getAvailableTokens(int $hitCount): int
```
Comment (Nyholm) [inline src/Symfony/Component/RateLimiter/SlidingWindowLimiter.php:100]: This method is public in `FixedWindow`, 

@wouterj is that a bug? It is not part of the interface and never used outside of the class.

File: src/Symfony/Component/RateLimiter/SlidingWindow.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+namespace Symfony\Component\RateLimiter;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 5.2
+ */
+final class SlidingWindow implements LimiterStateInterface
+{
+    private $id;
```
Comment (OskarStark) [inline src/Symfony/Component/RateLimiter/SlidingWindow.php:26]: Should we add phpdoc types for the properties used in this PR?

Comment (Nyholm) [reply src/Symfony/Component/RateLimiter/SlidingWindow.php:26]: I would love to. Yes

File: src/Symfony/Component/RateLimiter/SlidingWindowLimiter.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ * the current window. We have made 3 hits in current window.
+ * That means our sliding window hit count is (75% * 8) + 3 = 9.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 5.2
+ */
+final class SlidingWindowLimiter implements LimiterInterface
+{
+    private $id;
```
Comment (OskarStark) [inline src/Symfony/Component/RateLimiter/SlidingWindowLimiter.php:37]: Thanks for the docs above, would love to get some here too 😎

Comment (Nyholm) [reply src/Symfony/Component/RateLimiter/SlidingWindowLimiter.php:37]: Sure!

File: src/Symfony/Component/RateLimiter/SlidingWindow.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        return time() > $this->windowEndAt;
+    }
+
+    public function add(int $hits = 1)
+    {
+        $this->hitCount += $hits;
+    }
+
+    /**
+     * Calculate the sliding window number of request.
```
Comment (fabpot) [inline src/Symfony/Component/RateLimiter/SlidingWindow.php]: Calculates

Comment (Nyholm) [reply src/Symfony/Component/RateLimiter/SlidingWindow.php]: Thank you

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Nyholm.

**Comment by Nyholm** [comment]
Thank you for merging

**Pull request merged**